pkg/operator/controller/controller.go

```diff
@@ -266,11 +328,259 @@ func (r *ReconcileCDI) reconcileUpdate(logger logr.Logger, cr *cdiv1alpha1.CDI)
 		if err = r.ensureUploadProxyRouteExists(logger, cr); err != nil {
 			return reconcile.Result{}, err
 		}
+
+		if r.isUpgrading(cr) {
+			if err = r.cleanupUnusedResources(logger, cr); err != nil {
+				return reconcile.Result{}, err
+			}
+			previousVersion := cr.Status.ObservedVersion
+			cr.Status.ObservedVersion = r.namespacedArgs.DockerTag
+			if err = r.crUpdate(cdiv1alpha1.CDIPhaseDeployed, cr); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			logger.Info("Successfully finished upgrade from %v to %v and entered Deployed state", previousVersion, cr.Status.ObservedVersion)
+		}
 	}
 
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileCDI) cleanupUnusedResources(logger logr.Logger, cr *cdiv1alpha1.CDI) error {
+	//Iterate over installed resources of
+	//Deployment/CRDs/Services etc and delete all resources that
+	//do not exist in current version
+
+	targetStrategy, err := r.getAllResources(cr)
+	if err != nil {
+		return err
+	}
+
+	lo := &client.ListOptions{}
+	// maybe use different selectors?
+	lo.SetLabelSelector("cdi.kubevirt.io")
+
+	// remove unused CRDs
+	{
```

Comment:
There's got to be a way to use reflection/type info to reduce LOC here.  

Thinking in could start off like this:

```golang
for _, listType := range []runtime.Object{&corev1.ServiceList{}, &appsv1.DeploymentList{}, ...}

``` 

But need a function that will use reflection to call `.Items` and convert result to []runtime.Object.  Then can call GetName() and GetNamespace() on that.

This vexed me before (dealing with list types in a generic way) and I think this would be a fun thing to figure out.

Reply:
This code can be simplified quite a bit see here for inspiration:  https://github.com/kubevirt/containerized-data-importer/blob/24741566f3d8215356e8d9d8860f09a0373598f7/pkg/operator/controller/callbacks.go#L241-L268

Reply:
fixed